src/systems/filecoin_blockchain/vm/interpreter/vm_interpreter.go

```diff
@@ -39,131 +20,177 @@ func (vmi *VMInterpreter_I) ApplyMessageBatch(inTree st.StateTree, msgs []Messag
+	if err != nil {
+		return vmr.InvocOutput{
+			OutTree:     input.InTree,
+			GasUsed:     gascost.ApplyMessageFail,
+			ExitCode:    exitcode.ActorCodeNotFound,
+			ReturnValue: nil, // TODO: maybe: err
+		}
+	}
+
+	return actorCode.InvokeMethod(input, input.Method, input.Params)
 }
 
-func treeDeductFunds(inTree st.StateTree, a actor.Actor, amt actor.TokenAmount) (outTree st.StateTree) {
+func treeIncrementActorSeqNo(inTree st.StateTree, a actor.Actor) (outTree st.StateTree) {
 	panic("todo")
 }
 
-func treeDepositFunds(inTree st.StateTree, a actor.Actor, amt actor.TokenAmount) (outTree st.StateTree) {
+func treeDeductFunds(inTree st.StateTree, a actor.Actor, amt actor.TokenAmount) (outTree st.StateTree) {
+	// TODO: turn this into a single transfer call.
 	panic("todo")
 }
 
-func calcGas(gasUsed GasAmount, gasPrice GasPrice) {
-	return gasUsed * gasPrice
+func treeDepositFunds(inTree st.StateTree, a actor.Actor, amt actor.TokenAmount) (outTree st.StateTree) {
+	// TODO: turn this into a single transfer call.
+	panic("todo")
 }
 
-func treeGetOrCreateAccountActor(inTree st.StateTree, addr addr.Address) (outTree st.StateTree, _ actor.Actor) {
+func treeGetOrCreateAccountActor(inTree st.StateTree, a addr.Address) (outTree st.StateTree, _ actor.Actor, err error) {
 
-	toActor, found := inTree.GetActor(msg.To)
-	if found {
-		return inTree, toActor
+	toActor := inTree.GetActor(a)
+	if toActor != nil { // found
+		return inTree, toActor, nil
 	}
 
-	switch addr.Type() {
-	case BLS:
-		return newBLSAccountActor(inTree, addr)
-	case Secp256k1:
-		return newSecp256k1AccountActor(inTree, addr)
-	case ID:
-		Fatal("no actor with given ID")
-	case Actor:
-		Fatal("no such actor")
+	switch a.Type().Which() {
+	case addr.Address_Type_Case_BLS:
+		return treeNewBLSAccountActor(inTree, a)
+	case addr.Address_Type_Case_Secp256k1:
+		return treeNewSecp256k1AccountActor(inTree, a)
+	case addr.Address_Type_Case_ID:
+		return inTree, nil, errors.New("no actor with given ID")
+	case addr.Address_Type_Case_Actor:
+		return inTree, nil, errors.New("no such actor")
+	default:
+		return inTree, nil, errors.New("unknown address type")
 	}
 }
 
-func treeNewBLSAccountActor(inTree StateTree, addr Address) (outTree StateTree, _ Actor) {
+func treeNewBLSAccountActor(inTree st.StateTree, addr addr.Address) (outTree st.StateTree, _ actor.Actor, err error) {
 	panic("todo")
 }
 
-func treeNewSecp256k1AccountActor(inTree StateTree, addr Address) (outTree StateTree, _ Actor) {
+func treeNewSecp256k1AccountActor(inTree st.StateTree, addr addr.Address) (outTree st.StateTree, _ actor.Actor, err error) {
 	panic("todo")
 }
 
-//------------
+func applyError(tree st.StateTree, exitCode msg.ExitCode, gasUsed msg.GasAmount) (outTree st.StateTree, ret msg.MessageReceipt) {
+	return outTree, &msg.MessageReceipt_I{
+		ExitCode_:    exitCode,
+		ReturnValue_: nil,
+		GasUsed_:     gasUsed,
+	}
+}
+
+func gasToFIL(gas msg.GasAmount, price msg.GasPrice) actor.TokenAmount {
```

Comment:
why are these function not methods of some struct?

Reply:
they don't need to be. we need some simple helper functions. they do not need to be functions on structs